realm/realm-library/src/syncIntegrationTest/java/io/realm/SyncedRealmTests.java

```diff
@@ -0,0 +1,275 @@
+import io.realm.exceptions.RealmMigrationNeededException;
+import io.realm.objectserver.BaseIntegrationTest;
+import io.realm.objectserver.utils.Constants;
+import io.realm.rule.RunInLooperThread;
+import io.realm.rule.RunTestInLooperThread;
+import io.realm.rule.TestSyncConfigurationFactory;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+
+/**
+ * Catch all class for tests that not naturally fit anywhere else.
+ */
+public class SyncedRealmTests extends BaseIntegrationTest {
+
+    @Rule
+    public RunInLooperThread looperThread = new RunInLooperThread();
+
+    @Rule
+    public final UiThreadTestRule uiThreadTestRule = new UiThreadTestRule();
+
+    @Rule
+    public final TestSyncConfigurationFactory configurationFactory = new TestSyncConfigurationFactory();
+
+    @Test
+    @UiThreadTest
+    public void waitForInitialRemoteData_mainThreadThrows() {
+        final SyncUser user = loginUser();
+
+        SyncConfiguration config = new SyncConfiguration.Builder(user, Constants.USER_REALM)
+                .waitForInitialRemoteData()
+                .build();
+
+        Realm realm = null;
+        try {
+            realm = Realm.getInstance(config);
+            fail();
+        } catch (IllegalStateException ignored) {
+        } finally {
+            if (realm != null) {
+                realm.close();
+            }
+        }
+    }
+
+    // Login user on a worker thread, so this method can be used from both UI and non-ui threads.
+    @NonNull
+    private SyncUser loginUser() {
+        final CountDownLatch userReady = new CountDownLatch(1);
+        final AtomicReference<SyncUser> user = new AtomicReference<>();
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                SyncCredentials credentials = SyncCredentials.usernamePassword(UUID.randomUUID().toString(), "password", true);
+                user.set(SyncUser.login(credentials, Constants.AUTH_URL));
+                userReady.countDown();
+            }
+        }).start();
+        TestHelper.awaitOrFail(userReady);
+        return user.get();
+    }
+
+    @Test
+    public void waitForInitialRemoteData() {
+        // TODO We can improve this test once we got Sync Progress Notifications. Right now we cannot detect
+        // when a Realm has been uploaded.
+        SyncCredentials credentials = SyncCredentials.usernamePassword(UUID.randomUUID().toString(), "password", true);
+        SyncUser user = SyncUser.login(credentials, Constants.AUTH_URL);
+        SyncConfiguration config = new SyncConfiguration.Builder(user, Constants.USER_REALM)
+                .waitForInitialRemoteData()
+                .build();
+
+        Realm realm = null;
+        try {
+            realm = Realm.getInstance(config);
+            assertTrue(realm.isEmpty());
```

Comment:
Why empty? And what does this test?

I think there are no difference if we don't call `.waitForInitialRemoteData();`.